File: 002/python/main.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        root = ListNode()
        node = root
        r = 0
        while l1 is not None or l2 is not None or r != 0:
            r1 = 0 if l1 is None else l1.val
            r2 = 0 if l2 is None else l2.val
            r += r1 + r2
            
            val = r % 10
            node.next = ListNode(val)
            node = node.next
          
            l1 = None if l1 is None else l1.next
            l2 = None if l2 is None else l2.next
            r //= 10

        return root.next

File: 002/cpp/main.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *root = new ListNode();
        ListNode *temp = root;
        int r = 0;
        while (l1 != nullptr || l2 != nullptr || r != 0) {
            int a = l1 != nullptr ? l1->val : 0;
            int b = l2 != nullptr ? l2->val : 0;
            r += a + b;            
            temp->next = new ListNode(r % 10);
            temp = temp->next;
            l1 = l1 != nullptr ? l1->next : nullptr;
            l2 = l2 != nullptr ? l2->next : nullptr;
            r /= 10;
        }
        ListNode *first = root;
        root = root->next;
        delete first;
        return root;
    }
};
